pkg/cmd/cli/cmd/loginoptions.go

```diff
@@ -0,0 +1,325 @@
+
+			clientCfg.Insecure = util.PromptForBool(os.Stdin, "Use insecure connections [y/N]? ")
+			if !clientCfg.Insecure {
+				return err
+			}
+		}
+	}
+
+	// we have all info we need, now we can have a proper Config
+
+	o.Config = clientCfg
+
+	o.gatherServerInfo = true
+	return nil
+}
+
+// Negotiate a bearer token with the auth server, or try to reuse one based on the
+// information already present. In case of any missing information, ask for user input
+// (usually username and password, interactive depending on the Reader).
+func (o *LoginOptions) GatherAuthInfo() error {
+	o.assertGatheredServerInfo()
+
+	if me, err := whoami(o.Config); err == nil && (!o.usernameProvided() || (o.usernameProvided() && o.Username == usernameFromUser(me))) {
+		o.Username = usernameFromUser(me)
+		fmt.Printf("Already logged into '%v' as '%v'.\n", o.Config.Host, o.Username)
+
+	} else {
+		// if not, we need to log in again
+
+		o.Config.BearerToken = ""
+		_, err := tokencmd.RequestToken(o.Config, o.Reader, o.Username, o.Password)
+		if err != nil {
+			return err
+		}
+
+		me, err := whoami(o.Config)
+		if err != nil {
+			return err
+		}
+		o.Username = usernameFromUser(me)
+		fmt.Printf("Logged into '%v' as '%v'.\n", o.Config.Host, o.Username)
+	}
+
+	// if trying to log in an user that's not the currently logged in, try to reuse an existing token
+
+	// if o.usernameProvided() {
+	// 	glog.V(5).Infof("Checking existing authentication info for '%v'...\n", o.Username)
+
+	// 	for _, ctx := range rawCfg.Contexts {
+	// 		authInfo := rawCfg.AuthInfos[ctx.AuthInfo]
+	// 		clusterInfo := rawCfg.Clusters[ctx.Cluster]
+
+	// 		if ctx.AuthInfo == o.Username && clusterInfo.Server == o.Server && len(authInfo.Token) > 0 { // only token for now
+	// 			glog.V(5).Infof("Authentication exists for '%v' on '%v', trying to use it...\n", o.Server, o.Username)
+
+	// 			o.Config.BearerToken = authInfo.Token
+
+	// 			if me, err := whoami(o.Config); err == nil && usernameFromUser(me) == o.Username {
+	// 				o.Username = usernameFromUser(me)
+	// 				return nil
+	// 			}
+
+	// 			glog.V(5).Infof("Token %v no longer valid for '%v', can't use it\n", authInfo.Token, o.Username)
+	// 		}
+	// 	}
+	// }
+
+	o.gatherAuthInfo = true
+	return nil
+}
+
+// Discover the projects available for the stabilished session and take one to use. It
+// fails in case of no existing projects, and print out useful information in case of
+// multiple projects.
+func (o *LoginOptions) GatherProjectInfo() error {
+	o.assertGatheredAuthInfo()
+
+	oClient, err := client.New(o.Config)
+	if err != nil {
+		return err
+	}
+
+	projects, err := oClient.Projects().List(labels.Everything(), labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	projectsItems := projects.Items
+
+	switch len(projectsItems) {
+	case 0:
+		if me, err := whoami(o.Config); err == nil {
```

Comment:
TODO about touching up this message if we ever provide a server endpoint that allows an admin to describe to users how to request projects.